nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java

```diff
@@ -551,8 +551,52 @@ else if (!deleteFile(f) && undeletedFiles != null) {
      * @param f file to delete
      */
     public static void deleteFileNowOrLater(File f) {
-        if(!deleteFile(f))
-            f.deleteOnExit();
+        if(!deleteFile(f)) {
+            deleteOnExitRecursively(f);
+        }
+    }
+
+    /**
+     * Deletes the provided file by registering a shutdown hook with the
+     * Java Runtime. Non-empty directories will be deleted recursively.
+     * @param f the file to delete on JVM shutdown
+     */
+    public static void deleteOnExitRecursively(File f) {
+        if (f != null && f.exists()) {
+            final SecurityManager securityManager = System.getSecurityManager();
+            if (securityManager != null) {
+                securityManager.checkDelete(f.getAbsolutePath());
+            }
+            FILES_TO_DELETE_ON_SHUTDOWN.add(f);
```

Comment:
I wonder how much memory this will allocate and make unavailable for application. In some angle of view it could be considered as a memory leak.

Reply:
This mimics how to JDK does it. Except for the fact that the JDK holds a static list of Strings.